### PR TITLE
DM-33429: Add ability to do both serial and parallel overscan correction

### PR DESCRIPTION
**Pull request by czwa**
* Remove isrFunction intermediate for overscan correction.
* Add support for parallel overscan correction.
* Fix handling of column/row skipping, to ensure that overscan result is subtracted where needed.
* Consolidate unit tests for overscan correction in one location.

### CODE REVIEW

File: python/lsst/ip/isr/overscan.py [hunk 1]
```diff
@@ -244,7 +463,6 @@ def measureConstantOverscan(self, image):
         overscanValue = afwMath.makeStatistics(calcImage, fitType, self.statControl).getValue()
 
         return pipeBase.Struct(overscanValue=overscanValue,
-                               maskArray=None,
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:247]: Need to remove this from the docstring too.

File: python/lsst/ip/isr/overscan.py [hunk 2]
```diff
@@ -451,7 +670,7 @@ def maskExtrapolated(collapsed):
                 maskArray[-high:] = True
         return maskArray
 
-    def measureVectorOverscan(self, image):
+    def measureVectorOverscan(self, image, isTransposed=False):
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:686]: New parameter need to be documented.

File: python/lsst/ip/isr/overscan.py [hunk 3]
```diff
@@ -283,7 +501,8 @@ def transpose(imageArray):
         isTransposed : `bool`
             Indicates whether the input data was transposed.
         """
-        if np.argmin(imageArray.shape) == 0:
+        # if np.argmin(imageArray.shape) == 0:
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:517]: Remove commented code.

File: python/lsst/ip/isr/overscan.py [hunk 4]
```diff
@@ -268,7 +486,7 @@ def getImageArray(self, image):
         return calcImage
 
     @staticmethod
-    def transpose(imageArray):
+    def transpose(imageArray, isTransposed=False):
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:502]: This seems like a meaningless addition to me: telling the function whether you want it to do something. If you don't want it to do anything, don't call it in the first place.

File: tests/test_isrFunctions.py [hunk 5]
```diff
@@ -254,7 +254,7 @@ def test_illumCorrection(self):
         with self.assertRaises(RuntimeError):
             ipIsr.illuminationCorrection(self.mi, flatMi, 1.0, trimToFit=False)
 
-    def test_overscanCorrection_isInt(self):
+    def notest_overscanCorrection_isInt(self):
```
Comment (PaulPrice) [inline tests/test_isrFunctions.py]: Why have these been disabled? There's no note in the commit message, and it's not clear what "to check for Jenkins level failures" means.
Besides disabling the tests, there are other functional changes, which should probably be part of a different commit.

File: python/lsst/ip/isr/overscan.py [hunk 6]
```diff
@@ -93,12 +93,12 @@ class OverscanCorrectionTaskConfig(pexConfig.Config):
     )
     leadingRowsToSkip = pexConfig.Field(
         dtype=int,
-        doc="Number of leading columns to skip in serial overscan correction.",
+        doc="Number of leading rows to skip in parallel overscan correction.",
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:96]: Squash into previous commit.

File: python/lsst/ip/isr/overscan.py [hunk 7]
```diff
@@ -93,12 +93,12 @@ class OverscanCorrectionTaskConfig(pexConfig.Config):
...
         default=0,
     )
     trailingRowsToSkip = pexConfig.Field(
         dtype=int,
-        doc="Number of trailing columns to skip in serial overscan correction.",
+        doc="Number of trailing rows to skip in parallel overscan correction.",
```
Comment (PaulPrice) [inline python/lsst/ip/isr/overscan.py:101]: Ditto.

File: tests/test_overscanCorrection.py [hunk 8]
```diff
@@ -268,6 +305,115 @@ def notest_SplineOverscanCorrection(self):
...
+        oscanResults = self.task.overscanCorrection(self.inputExp, self.amp)
+        self.assertIsInstance(oscanResults, Struct)
+        self.assertIsInstance(oscanResults.imageFit, afwImage.ImageF)
+        self.assertIsInstance(oscanResults.overscanFit, afwImage.ImageF)
+        self.assertIsInstance(oscanResults.overscanImage, afwImage.MaskedImageF)
+
+        statAfter = computeImageMedianAndStd(self.inputExp.image[self.amp.getRawDataBBox()])
+        self.assertLess(statAfter[0], statBefore[0])
+
+    def notest_overscanCorrection_isInt(self):
```
Comment (PaulPrice) [inline tests/test_overscanCorrection.py]: These tests are still disabled.

File: python/lsst/ip/isr/isrTask.py [hunk 9]
```diff
@@ -443,48 +443,12 @@ class IsrTaskConfig(pipeBase.PipelineTaskConfig,
...
-    )
-    overscanNumSigmaClip = pexConfig.Field(
-        dtype=float,
-        doc="Rejection threshold (sigma) for collapsing overscan before fit",
-        default=3.0,
-        deprecated=("Please configure overscan via the OverscanCorrectionConfig interface."
-                    " This option will no longer be used, and will be removed after v20.")
-    )
-    overscanIsInt = pexConfig.Field(
+    doParallelOverscan = pexConfig.Field(
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/isrTask.py:446]: Agreed.  The new config option was a workaround for testing that can now be dropped (like the other deprecated options).

File: python/lsst/ip/isr/overscan.py [hunk 10]
```diff
@@ -74,6 +75,39 @@ class OverscanCorrectionTaskConfig(pexConfig.Config):
...
+    )
+    trailingRowsToSkip = pexConfig.Field(
+        dtype=int,
+        doc="Number of trailing columns to skip in serial overscan correction.",
+        default=0,
+    )
+
+    maxDeviation = pexConfig.Field(
+        dtype=float,
+        doc="Maximum deviation from median to mask in overscan correction.",
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/overscan.py:107]: These should be ADU (docstring updated to reflect that).  This value is also much larger than the expected read noise (~200x), that I think keeping it in ADU makes it clear that this is just attempting to remove any bad edges or CR hits that are unreasonably bright.

File: python/lsst/ip/isr/overscan.py [hunk 11]
```diff
@@ -101,17 +135,17 @@ def __init__(self, statControl=None, **kwargs):
...
-            Image data that will have the overscan removed.
-        overscanImage : `lsst.afw.image.Image`
-            Overscan data that the overscan is measured from.
-        amp : `lsst.afw.cameraGeom.Amplifier`, optional
+        exposure : `lsst.afw.image.Exposure`
+            Image data that will have the overscan corrections applied.
+        amp : `lsst.afw.cameraGeom.Amplifier`
             Amplifier to use for debugging purposes.
+        kwargs : Any
+            Additional parameters.
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/overscan.py:151]: Removed as it was supporting the config workaround discussed above.

File: python/lsst/ip/isr/overscan.py [hunk 12]
```diff
@@ -134,55 +168,240 @@ def run(self, ampImage, overscanImage, amp=None):
         ------
         RuntimeError
             Raised if an invalid overscan type is set.
+        """
+        # Config overrides.
+        leadingColumnsToSkip = kwargs['leadingColumnsToSkip'] if 'leadingColumnsToSkip' in kwargs \
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/overscan.py:176]: Also now removed entirely.

File: python/lsst/ip/isr/overscan.py (hunk 4 above)
Comment (czwa) [reply python/lsst/ip/isr/overscan.py:502]: I've dropped the method, as it's trivial without that logic.

File: python/lsst/ip/isr/overscan.py [hunk 13]
```diff
@@ -134,55 +168,240 @@ def run(self, ampImage, overscanImage, amp=None):
...
+        -------
+        overscanModel : `np.ndarray`, (Nrows, Ncols) or scalar
+            Expanded overscan fit.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if no axis has the appropriate dimension.
         """
+        if type(overscanValue) == np.ndarray:
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/overscan.py:320]: I've played around with `broadcast_to`, but I don't think it easily handles all the cases (my test code is requiring `atleast_2d` and `transpose` calls that I'm not sure make this simpler).

File: python/lsst/ip/isr/overscan.py [hunk 14]
```diff
@@ -134,55 +168,240 @@ def run(self, ampImage, overscanImage, amp=None):
...
+        overscanMean = serialResults.overscanMean
+        overscanSigma = serialResults.overscanSigma
+
+        stats = afwMath.makeStatistics(overscanImage.getImage(),
+                                       afwMath.MEDIAN | afwMath.STDEVCLIP, self.statControl)
+        residualMean = stats.getValue(afwMath.MEDIAN)
+        residualSigma = stats.getValue(afwMath.STDEVCLIP)
+
+        print(f"CZW: M: {overscanMean} S: {overscanSigma} RM: {residualMean} RS: {residualSigma}")
+        # CZW: Do we ever do this?
```
[the remark this answers is not in the document]
Comment (czwa) [reply python/lsst/ip/isr/overscan.py]: I've removed the code here, as it was doing nothing.

File: tests/test_isrFunctions.py (hunk 5 above)
Comment (czwa) [reply tests/test_isrFunctions.py]: This has been removed, as the coverage is handled by the `test_overscanCorrection.py` tests.

File: tests/test_overscanCorrection.py (hunk 8 above)
Comment (czwa) [reply tests/test_overscanCorrection.py]: Coverage now provided by other tests and this method removed.

**Pull request merged**